### ISSUE TITLE
UX for time-correlated metrics

### ISSUE THREAD

**Issue by pasko** [catapult-project/catapult#4161]
## Background

The http://crbug.com/796957#c18 describes one confusion when using the "View trace from this run" functionality on the chromeperf dashboard: the value on the graph on the dashboard does not closely match the one reported in the corresponding trace. Let me try to explain what is happening and what we can do about it with multiple teams involved :)

## Root cause

The process seems to work as follows:
* The bot runs the benchmark 10 times
* The first 2 runs are a lot slower than the subsequent ones (let's refer to the metric as 'metric')
* The dashboard displays the 'metric_avg', which is very close in value to what is recorded on the runs 3-10
* The link to "View trace from this run" directs to the first trace (i.e. the most non-representative one)

###  Make the benchmarks responsible for warming up and removing dependence on iteration number

The advantages are that the infrastructure above the benchmark becomes automatically easy and intuitive. Unfortunately, for some benchmarks it means slower times to run, potentially slower or broken bisects (because with overridden `--pageset-repeat` it is not generally clear how to run the warmup). Is there an API in telemetry to define how many iterations to run for warmup?

### Display more information on the dashboard

The dashboard could add links to all traces from the run. Also it could display all of the contents of the histogram, not just the average (there seem to exist some plans around this, Pinpoint should greatly appreciate the increased statistical power). This may make it less intuitive to relate what points are coming from a single run, not sure how important it is.

If the benchmarks are have weird time-dependent effects, it also adds complexity to Pinpoint, since changing the value of `--pageset-repeat` adds bias to metrics.

## Related UX problems noticed

* The link to "Buildbot stdio" is broken
* The link to "Download as CSV" is well hidden and reports only one value per revision, effectively dropping 9/10 of statistical information

**Comment by nedn** [comment]
@benshayden @dave-2 @eakuefner

**Comment by benshayden** [comment]
Thanks for the bug report! And apologies for the delay as I ramp back up from leave.

> Make the benchmarks responsible for warming up and removing dependence on iteration number
> Is there an API in telemetry to define how many iterations to run for warmup?

@nedn Can you address this question?

> Display more information on the dashboard

The Histogram Pipeline and V2SPA projects aim to do just this.

 * [TBMv2](https://github.com/catapult-project/catapult/blob/master/docs/how-to-write-metrics.md) introduced a new system of abstractions for metrics and benchmarks to record two orders of magnitude more information in Histograms and Diagnostics. The dashboard currently ingests chartjson via /add_point. The Histogram Pipeline project is updating the dashboard to ingest Histograms with all their Diagnostics. For example, chartjson records a single trace per story (not story run), but Histograms will record all traces as well as other artifacts, and link them to the individual samples within the Histograms for easy access.
 * V2SPA is a ground-up rewrite of the dashboard frontend as a single-page app to display Histograms and Diagnostics, simplify important work flows, and update the UI to be fully Material Design. [[prototype](v2spa-dot-chromeperf.appspot.com/spa)] I'll share the design doc with you as soon as it's ready.

> The link to "Buildbot stdio" is broken

Can you link to a chart with a broken link? I looked in [796957#c18](http://crbug.com/796957#c18), found [messageloop_start_time](https://chromeperf.appspot.com/report?sid=15c2861109c546f72304a81d85a3692578ce2256e4b012df32602fa8f7b72089&start_rev=1515607362&end_rev=1516148025), but all the Buildbot stdio links on that chart work for me. You may need to sign into a google.com account.

> The link to "Download as CSV" is well hidden and reports only one value per revision

I'll make a note to follow up with you when I've prototyped Export buttons in V2SPA.

**Comment by pasko** [comment]
Here is a [chart with a broken link](https://chromeperf.appspot.com/report?sid=9df0ec76b0fffc74931cc7bdcf44cd4d540f81bb34ccde3571c2211ec8360ee6).

When I click on a point, say "Point ID: 533259", and then "Buildbot stdio", it leads to [this log](https://luci-logdog.appspot.com/v/?s=chrome%2Fbb%2Fchromium.perf.fyi%2FAndroid_Nexus_5X_Perf_FYI%2F2199%2F%2B%2Frecipes%2Fsteps%2Fexperimental.startup.android.coldish%2F0%2Fstdout). The log says "(Missing)". I am viewing both chromeperf and luci-logdog while signed in with @google.com.

**Comment by nedn** [comment]
> Make the benchmarks responsible for warming up and removing dependence on iteration number
Is there an API in telemetry to define how many iterations to run for warmup?

We have to add param this to start up benchmark itself, there is not core support for this. We can take this into account when designing the newer startup benchmark.

@perezju

**Comment by pasko** [comment]
Thank you for the update, @benshayden. I heard about the effort to publish more histogram data, eager to start using it, feel free to create a separate issue for it, if you haven't already.

The TBMv1 benchmark overloaded the ValueCanBeAddedPredicate() to ignore the first run. If something like this gets exposed on the metrics side, it may work as well.

**Comment by perezju** [comment]
Could this be solved at the [metric](https://cs.chromium.org/chromium/src/third_party/catapult/tracing/tracing/metrics/) level?

On each story the startup benchmark runs the flow of "launch browser, open page, close browser" twice (or as many times the benchmark deems it's needed). We get a single trace for this *one* story which will include two (or more) startup measurements.

The metric can then read both of those from the trace, report the first as "cold" and the later one as "warm" measurement.

Keeping both browser startup events in the same trace might also help to debug in the case of e.g. regressions.

Any thoughts?

**Comment by pasko** [comment]
I was not sure one trace for multiple startups is possible. It may simplify things!

Does it imply merging traces that we pull from devtools or some other mechanism? Another possibility is that tracing can be configured to dump the trace to disk under a certain name as soon as some certain event appears. There are pros and cons in each. What if systrace needs to be enabled, will the trace merging be affected?

But besides that, I am hesitant also because a trace across 30 runs would make it a ~150 MiB html file, and may hit various limits.

**Comment by nedn** [comment]
@pasko, Telemetry TBMv2 automatically merge all the traces that it collected, whether they are multiple chrome traces, or multiple chrome traces along with systrace.

For the trace size, we have a streaming algorithm in parsing the trace that helped mitigate the risk of OOM. Also the memory limit of trace analysis is[ set to 8Gb](https://cs.chromium.org/chromium/src/third_party/catapult/tracing/tracing/mre/map_single_trace.py?rcl=12ffa89339910751ca0516770d7b8877f4dd25a3&l=103), so it's generally safe.

**Comment by pasko** [comment]
nice! I was confused because in the '*.coldish' benchmark I saw several traces uploaded, and never saw the merged trace.

**Comment by perezju** [comment]
> But besides that, I am hesitant also because a trace across 30 runs would make it a ~150 MiB html file

Do we really need to repeat each "unit" (launch/load/close) 30 times? How about just doing 2 or 3 launches per story and then, to de-noise, run with e.g. `--pageset-repeat 10`?

This is also consistent with the way that bisect wants to look at things, and will repeat a story (in this case one with 2 or 3 launches in it) enough times until it gets a discriminating signal.

**Comment by pasko** [comment]
I was not thinking about bisect, nor I worried about regular benchmark runs. We currently set --pageset-repeat=11 for startup benchmarks (when not under bisect), no reason to increase it to 30 yet.

Just imagining a more in-depth statistical analysis being done locally. This reminds me that we made some studies with ~1000 runs (for customtabs_benchmark and some startup), and it was nice to be able to just leave the device running and logging the data (pulled into ipython notebook later). For such a study we would not need to look at the trace all at once anyway, and 8GiB should be enough for everyone.

To summarize: I'd like to keep in mind the ability to increase --pageset-repeat to large numbers, but it is not affecting us in the context of memory limitations, it seems.

**Comment by perezju** [comment]
Yes. To be clear what I suggest is to decouple `--pageset-repeat n`  from the number of "launch/load/close" cycles in your benchmark.

If you say `--pageset-repeat 1000` you will always get 1000 individual traces for each story in your benchmark.

Except that now each one of those traces will contain 2 or 3 "launch/load/close" cycles with both cold and warm measurements as defined by your benchmark.

**Comment by pasko** [comment]
Ah, this decoupling sounds nice. I guess the only disadvantage is that when bisecting/analysing a warm regression/codechange Pinpoint/etc would spend time running cold starts as well. Or perhaps it needs applying the story filter art.. Anything else to worry about?

**Comment by perezju** [comment]
> I guess the only disadvantage is that when bisecting/analysing a warm regression/codechange Pinpoint/etc would spend time running cold starts as well.

I wouldn't worry too much about that, and it would certainly be better than the current state where by mucking with `--pageset-repeat` we may break the semantics of your metrics.

**Comment by benshayden** [comment]
Archiving. Please file a new bug in crbug if changes are required in the chromeperf dashboard.
